File: src/domain/palette.js
// Colours that tell one player from the next.
//
// Colour encodes the PLAYER, never the outcome -- outcomes are carried by the shape of a
// mark, so the board stays readable without colour vision (FR-034).
//
// It used to encode the turn. "The green tallies are number 5" is something a coach can
// hold in their head across a whole match; "the green ones are the third turn of this
// match" is not. The native app changed first; this is the same rule, computed the same
// way, so a season looks the same in both.

// Six hues, ordered so that neighbouring entries are far apart on the colour wheel and so
// that the wrap from the last back to the first is far apart too. Every one clears WCAG AA
// contrast against the app's dark background.
export const PALETTE = Object.freeze([
  '#22d3ee', // cyan
  '#f59e0b', // amber
  '#818cf8', // indigo
  '#a3e635', // lime
  '#fb7185', // rose
  '#e879f9', // fuchsia
])

// The same six as hue, saturation and lightness, because a player's own turns are shades of
// their hue and a shade is a change to one of these numbers.
export const PLAYER_PALETTE = Object.freeze([
  Object.freeze({ hue: 188, saturation: 0.86, lightness: 0.53 }),
  Object.freeze({ hue: 38, saturation: 0.92, lightness: 0.5 }),
  Object.freeze({ hue: 234, saturation: 0.89, lightness: 0.74 }),
  Object.freeze({ hue: 83, saturation: 0.78, lightness: 0.55 }),
  Object.freeze({ hue: 351, saturation: 0.95, lightness: 0.71 }),
  Object.freeze({ hue: 292, saturation: 0.92, lightness: 0.73 }),
])

// How a player's own turns are told apart: lighter, darker, lighter still. Applied to the
// lightness, never the hue, so the colour still reads as theirs from across a court.
export const TURN_SHADES = Object.freeze([0, 0.14, -0.11, 0.25])

// The band of lightness that stays legible on the dark board at both ends.
const LIGHTEST = 0.88
const DARKEST = 0.36

/**
 * The colour for a serve turn at the given position within its match.
 *
 * Kept because the reducer stamps `colorIndex` on every turn from it, and that index is
 * part of the state a replayed log produces -- changing it would change what an old log
 * replays to, which is the one thing a migration must never do.
 */
export function colorForTurn(ordinal) {
  return PALETTE[colorIndexForTurn(ordinal)]
}

/** The palette index for a turn ordinal. Adjacent ordinals never collide. */
export function colorIndexForTurn(ordinal) {
  return ((ordinal % PALETTE.length) + PALETTE.length) % PALETTE.length
}

/**
 * The palette index for a player, by their place in the order they first served.
 *
 * Wrapping is unavoidable with more players than hues; the seventh takes the first hue
 * again, which is why `colorForPlayer` also darkens on each lap round.
 */
export function colorIndexForPlayer(index) {
  const count = PLAYER_PALETTE.length
  return ((index % count) + count) % count
}

/** A player's own colour, at their base shade. */
export function colorForPlayer(index) {
  return colorForPlayerTurn(index, 0)
}

/**
 * The colour of one turn: the player's hue, at the shade that turn of theirs takes.
 *
 * `turnIndex` counts the player's own turns, not the match's -- a player's second turn is
 * their second shade whether it was the second serve turn of the match or the ninth.
 */
export function colorForPlayerTurn(playerIndex, turnIndex) {
  const base = PLAYER_PALETTE[colorIndexForPlayer(playerIndex)]
  const shades = TURN_SHADES
  const shade = shades[((turnIndex % shades.length) + shades.length) % shades.length]

  // Past the sixth player the hues start again, so each lap round is darkened to keep the
  // seventh player from being mistaken for the first.
  const lap = playerIndex < 0 ? 0 : Math.floor(playerIndex / PLAYER_PALETTE.length)
  const lightness = clamp(base.lightness + shade - lap * 0.18, DARKEST, LIGHTEST)

  return hslToHex(base.hue, base.saturation, lightness)
}

/**
 * HSL to the hex the rest of the app draws in.
 *
 * Written out rather than left to CSS because the Swift app computes the same colours from
 * the same numbers, and a season has to look the same in both.
 */
export function hslToHex(hue, saturation, lightness) {
  const chroma = (1 - Math.abs(2 * lightness - 1)) * saturation
  const sector = (((hue % 360) + 360) % 360) / 60
  const second = chroma * (1 - Math.abs((sector % 2) - 1))
  const base = lightness - chroma / 2

  const [red, green, blue] =
    sector < 1 ? [chroma, second, 0]
    : sector < 2 ? [second, chroma, 0]
    : sector < 3 ? [0, chroma, second]
    : sector < 4 ? [0, second, chroma]
    : sector < 5 ? [second, 0, chroma]
    : [chroma, 0, second]

  return `#${[red, green, blue].map((channel) => hex(channel + base)).join('')}`
}

function hex(value) {
  return Math.round(clamp(value, 0, 1) * 255).toString(16).padStart(2, '0')
}

function clamp(value, low, high) {
  return Math.min(Math.max(value, low), high)
}

File: packages/VBCore/Sources/VBCore/Palette.swift
// Colours that tell one serve turn from the next.
//
// Colour encodes the TURN, never the outcome. Outcomes are carried by the shape of a mark,
// so the tally board stays readable without colour vision.
import Foundation

/// Six hues, ordered so neighbouring entries are far apart on the colour wheel -- and so
/// the wrap from the last back to the first is far apart too. Every one clears WCAG AA
/// contrast against the app's dark background.
public let turnPalette = [
    "#22d3ee",  // cyan
    "#f59e0b",  // amber
    "#818cf8",  // indigo
    "#a3e635",  // lime
    "#fb7185",  // rose
    "#e879f9",  // fuchsia
]

/// The palette index for a turn ordinal. Adjacent ordinals never collide.
///
/// Depends only on the ordinal, so a turn's colour never changes because of something that
/// happened later -- which is what keeps replay deterministic.
public func colorIndexForTurn(_ ordinal: Int) -> Int {
    let count = turnPalette.count
    return ((ordinal % count) + count) % count
}

/// The colour for a serve turn at the given position within its match.
public func colorForTurn(_ ordinal: Int) -> String {
    turnPalette[colorIndexForTurn(ordinal)]
}

// MARK: - A colour per player

/// One hue per player, so a colour means a person rather than a turn.
///
/// "The green tallies are number 5" is a thing a coach can hold in their head across a whole
/// match; "the green ones are the third turn of this match" is not. These are the same six
/// hues the turn palette uses, kept because their contrast against the dark board was
/// already worked out -- what changes is what they are attached to.
///
/// Hue, saturation and lightness rather than hex, because turns within a player are shades
/// of their hue, and a shade is a change to one of these numbers.
public let playerPalette: [(hue: Double, saturation: Double, lightness: Double)] = [
    (188, 0.86, 0.53),  // cyan
    (38, 0.92, 0.50),  // amber
    (234, 0.89, 0.74),  // indigo
    (83, 0.78, 0.55),  // lime
    (351, 0.95, 0.71),  // rose
    (292, 0.92, 0.73),  // fuchsia
]

/// How a player's own turns are told apart: lighter, darker, lighter still.
///
/// Applied to the player's lightness, never their hue, so the colour still reads as theirs
/// from across a court. Four steps because a player rarely serves more than four turns in a
/// match, and a fifth repeating the first is no worse than two turns being the same shade.
public let turnShades: [Double] = [0, 0.14, -0.11, 0.25]

/// The band of lightness that stays legible on the dark board at both ends.
private let legibleLightness = 0.36...0.88

/// The palette index for a player, by their place in the order they first served.
///
/// Wrapping is unavoidable with more players than hues; the seventh player takes the first
/// hue again, which is why `colorForPlayer` also darkens on each lap round.
public func colorIndexForPlayer(_ index: Int) -> Int {
    let count = playerPalette.count
    return ((index % count) + count) % count
}

/// A player's own colour, at their base shade.
public func colorForPlayer(_ index: Int) -> String {
    colorForTurn(playerIndex: index, turnIndex: 0)
}

/// The colour of one turn: the player's hue, at the shade that turn of theirs takes.
///
/// `turnIndex` counts the player's own turns, not the match's -- a player's second turn is
/// their second shade whether it was the second serve turn of the match or the ninth.
public func colorForTurn(playerIndex: Int, turnIndex: Int) -> String {
    let base = playerPalette[colorIndexForPlayer(playerIndex)]
    let shade = turnShades[((turnIndex % turnShades.count) + turnShades.count) % turnShades.count]

    // Past the sixth player the hues start again, so each lap round is darkened to keep the
    // seventh player from being mistaken for the first.
    let lap = Double(max(0, index(ofLap: playerIndex)))
    let lightness = (base.lightness + shade - lap * 0.18).clamped(to: legibleLightness)

    return hex(hue: base.hue, saturation: base.saturation, lightness: lightness)
}

/// Which time round the palette this player is.
private func index(ofLap playerIndex: Int) -> Int {
    playerIndex < 0 ? 0 : playerIndex / playerPalette.count
}

/// HSL to the hex the rest of the app draws in.
///
/// Written out rather than pulled from a colour library because the package has no
/// dependencies, and because a colour that cannot be computed on this workstation cannot be
/// tested on it either.
public func hex(hue: Double, saturation: Double, lightness: Double) -> String {
    let chroma = (1 - abs(2 * lightness - 1)) * saturation
    let sector = (hue.truncatingRemainder(dividingBy: 360) + 360).truncatingRemainder(dividingBy: 360) / 60
    let second = chroma * (1 - abs(sector.truncatingRemainder(dividingBy: 2) - 1))
    let base = lightness - chroma / 2

    let (red, green, blue): (Double, Double, Double) =
        switch sector {
        case ..<1: (chroma, second, 0)
        case ..<2: (second, chroma, 0)
        case ..<3: (0, chroma, second)
        case ..<4: (0, second, chroma)
        case ..<5: (second, 0, chroma)
        default: (chroma, 0, second)
        }

    let channels = [red, green, blue].map { UInt8(((($0 + base) * 255).rounded()).clamped(to: 0...255)) }
    return "#" + channels.map { String(format: "%02x", $0) }.joined()
}

extension Double {
    /// Keeps a value inside a range, so a shade can be nudged without leaving the band that
    /// stays readable.
    fileprivate func clamped(to range: ClosedRange<Double>) -> Double {
        Swift.min(Swift.max(self, range.lowerBound), range.upperBound)
    }
}
